src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/ReadsPreprocessingPipeline.java

```diff
@@ -83,14 +88,27 @@ protected void setupPipeline( Pipeline pipeline ) {
                     : IntervalUtils.getAllIntervalsForReference(readsHeader.getSequenceDictionary());
 
             final PCollectionView<SAMFileHeader> headerSingleton = ReadsDataflowSource.getHeaderView(pipeline, readsHeader);
-            final PCollection<GATKRead> initialReads = readsDataflowSource.getReadPCollection(intervals);
 
             final OpticalDuplicateFinder finder = opticalDuplicatesArgumentCollection.READ_NAME_REGEX != null ?
                     new OpticalDuplicateFinder(opticalDuplicatesArgumentCollection.READ_NAME_REGEX, opticalDuplicatesArgumentCollection.OPTICAL_DUPLICATE_PIXEL_DISTANCE, null) : null;
             final PCollectionView<OpticalDuplicateFinder> finderPcolView = pipeline.apply(Create.of(finder)).apply(View.<OpticalDuplicateFinder>asSingleton());
 
-            // Apply MarkDuplicates to produce updated GATKReads.
-            final PCollection<GATKRead> markedReads = initialReads.apply(new MarkDuplicates(headerSingleton, finderPcolView));
+            // we read from shards that span this many bp
+            int basesPerShard = 1_000_000;
+            // and then output shards that span this many
+            int outputBasesPerWorkUnit = 5_000;
+
+            PCollection<GATKRead> markedReads;
+
+            if (oldMarkDups) {
+                // Apply MarkDuplicates to produce updated GATKReads.
+                final PCollection<GATKRead> initialReads = readsDataflowSource.getReadPCollection(intervals);
+                markedReads = initialReads.apply(new MarkDuplicates(headerSingleton, finderPcolView));
+            } else {
+                // Optimized MarkDuplicates
+                final PCollection<KV<String, Iterable<GATKRead>>> readsByShard = readsDataflowSource.getGroupedReadPCollection(intervals, basesPerShard, outputBasesPerWorkUnit, pipeline);
+                markedReads = readsByShard.apply(new MarkDuplicatesFromShardsDataflowTransform(headerSingleton, finderPcolView));
```

Comment:
Can you follow the convention (here and elsewhere in this branch) of naming the optimized version of an existing class X "OptimizedX" (eg., here `OptimizedMarkDuplicates`). This will make it a bit less confusing to wade through the different versions of tools/transforms.